config/src/bloop/config/Config.scala

```diff
@@ -164,10 +164,20 @@ object Config {
     val All: List[String] = List(NoModule.id, CommonJSModule.id, ESModule.id)
   }
 
+  sealed abstract class ModuleSplitStyleJS(val id: String)
+  object ModuleSplitStyleJS {
+    case object FewestModules extends ModuleSplitStyleJS("FewestModules")
+    case object SmallestModules extends ModuleSplitStyleJS("SmallestModules")
+    case object SmallModulesFor extends ModuleSplitStyleJS("SmallModulesFor")
+    val All: List[String] =
+      List(FewestModules.id, SmallestModules.id, SmallModulesFor.id)
+  }
+
   case class JsConfig(
       version: String,
       mode: LinkerMode,
       kind: ModuleKindJS,
+      moduleSplitStyle: Option[ModuleSplitStyleJS],
```

Comment:
So one thing that causes a lot of issues is that we'll have release 3.x versions after this and any other changes will require another major bump. The big issue here is case classes.

I am thinking of dropping case classes so that it's easier for us to add new stuff. What do you think?  

Anyone has ideas about what could be a good replacement?

Reply:
Another possibility for now would be to instead add a field `jsOptions` that would be a `List[JsOption]` and have all the options implement that. At least this would limit the need to add new js options as fields.

Reply:
Thanks! That looks interesting, but I wonder how it maintains the binary compatibility of unapply methods :thinking: It doesn't seem possible to easily do.

Reply:
Anyway, it seems that this should work @KristianAN would you want to take a look at that or can I modify the PR?

Reply:
@tgodzik I can take a look at how we can maintain binary compatibility.